include/zio/zio_dev.h

```diff
@@ -0,0 +1,342 @@
+	 */
+	const u8_t bit_width;
+
+	/**
+	 * Byte size of channel data, ex: 2 bytes (s16_t)
+	 */
+	const u8_t byte_size;
+
+	/**
+	 * Byte ordering of channel data
+	 */
+	const enum byteorder {
+		big_endian,
+		little_endian,
+	} byte_order;
+};
+
+/**
+ * @brief Function to set a device attribute
+ *
+ * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ */
+typedef int (*zio_dev_set_attr_t)(struct device *dev, u32_t attr_idx,
+				  const struct zio_attr_data *attr);
+
+/**
+ * @brief Function to get a device attribute
+ *
+ * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ */
+typedef int (*zio_dev_get_attr_t)(struct device *dev, uint32_t attr_idx,
+				  struct zio_attr_data *attr);
+
+
+/**
+ * @brief Function to get a pointer to the array of attribute descriptions
+ */
+typedef int (*zio_dev_get_attrs_t)(struct device *dev,
+				   struct zio_dev_attr *attrs,
+				   u16_t *num_attrs);
+
+/**
+ * @brief Function to get a pointer to the array of channel descriptions
+ */
+typedef int (*zio_dev_get_chans_t)(struct device *dev,
+				   struct zio_dev_chan *chans,
+				   u16_t *num_chans);
+
+/**
+ * @brief Function to get a channels attribute descriptions
+ *
+ * @return -EINVAL if an invalid channel id was given, 0 otherwise
+ */
+typedef int (*zio_chan_get_attrs_t)(struct device *dev, const u32_t chan_idx,
+				    struct zio_chan_attr *attrs,
+				    u16_t *num_attrs);
+
+/**
+ * @brief Function to set a channel attribute for a device
+ *
+ * @return -EINVAL if an invalid attribute id, type, or range was given,
+ * 0 otherwise
+ */
+typedef int (*zio_chan_set_attr_t)(struct device *dev, const u32_t chan_idx,
+				   const u32_t attr_idx,
+				   const struct zio_attr_data attr);
+/**
+ * @brief Function to get a channel attribute for a device
+ *
+ * @return -EINVAL if an invalid attribute id was given, 0 otherwise
+ */
+typedef int (*zio_chan_get_attr_t)(struct device *dev, const u32_t chan_idx,
+					u32_t attr_idx,
+					struct zio_attr_data *attr);
+
+/**
+ * @brief Function to enable a channel for a device
+ *
+ * Not all devices support enabling/disabling channels
+ *
+ * @return -EINVAL if an invalid chan was given, 0 otherwise
+ */
+typedef int (*zio_chan_enable_t)(struct device *dev, u32_t chan_idx);
+
+/**
+ * @brief Function to disable a channel for a device
+ *
+ * Not all devices support enabling/disabling channels
+ *
+ * @return -EINVAL if an invalid chan was given, 0 otherwise
+ */
+typedef int (*zio_chan_disable_t)(struct device *dev, u32_t chan_idx);
+
+/**
+ * @brief Function to determine if a channel for a device is enabled
+ *
+ * Not all devices support enabling/disabling channels
+ *
+ * @return -EINVAL if an invalid chan was given, 0 if not enabled, 1 if enabled
+ */
+typedef int (*zio_chan_is_enabled_t)(struct device *dev,
+				     struct zio_dev_chan *chan);
+
+
+/**
+ * @brief Function to trigger a read or write of the device
```

Comment:
Does it block until the read is finished?

Reply:
It would not block. The same buffer polling and pulling function pair would be used to read values back. For ad-hoc reading and writing a function would wrap around the idea of a manual trigger, poll, and pull

Reply:
I wouldn't expect it to block, but it's worth documenting.